### PR TITLE
Improve checking of `in` operator combined with `noUncheckedIndexedAccess`

### PR DESCRIPTION
**Pull request by Andarist**
follow up to https://github.com/microsoft/TypeScript/pull/50666 , mostly piggy-backing on the logic introduced there

fixes https://github.com/microsoft/TypeScript/issues/43614

cc @ahejlsberg @andreialecu

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -25251,17 +25251,33 @@ namespace ts {
...
+                    // `{ prop?: T }[prop]` always implies `T | undefined`
+                    if (!compilerOptions.noUncheckedIndexedAccess) {
+                        return filtered;
+                    }
+                    return mapType(filtered, t => {
+                        if (getPropertyOfType(type, name)) {
+                            return t;
+                        }
+                        return intersectWithNarrowingRecordType(type, nameType, getApplicableIndexInfoForName(type, name)!.type);
+                    });
```
Comment (Andarist) [inline src/compiler/checker.ts:25266]: I probably also need to handle optional properties here too - or at least add some tests for them.

Comment (Andarist) [reply src/compiler/checker.ts:25266]: Oh, I've actually conflated `exactOptionalPropertyTypes` with `noUncheckedIndexedAccess` somehow and I've thought that optional properties would also require some work here. But `noUncheckedIndexedAccess` doesn't affect optional properties anyhow - so the added code looks OK to me now. Also - I've rechecked the current behavior with `exactOptionalPropertyTypes` and without it and it looks OK.

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -25251,17 +25251,33 @@ namespace ts {
                 if (isKnownProperty) {
                     // If the check is for a known property (i.e. a property declared in some constituent of
                     // the target type), we filter the target type by presence of absence of the property.
-                    return filterType(type, t => isTypePresencePossible(t, name, assumeTrue));
+                    const filtered = filterType(type, t => isTypePresencePossible(t, name, assumeTrue));
+                    // without `noUncheckedIndexedAccess` we can't narrow any further
+                    // `{ [x: PropertyKey]: T }[prop]` already returns `T`
+                    // `{ prop?: T }[prop]` always implies `T | undefined`
```
Comment (Andarist) [inline src/compiler/checker.ts:25257]:
```suggestion
```

**Comment by JakobBruenker** [comment]
Thanks! This is exactly what's preventing us from using `noUncheckedIndexedAccess`, so I'm looking forward to seeing this merged

File: src/compiler/checker.ts [hunk 3]
```diff
@@ -25500,17 +25500,32 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
...
                 // If the check is for a known property (i.e. a property declared in some constituent of
                 // the target type), we filter the target type by presence of absence of the property.
-                return filterType(type, t => isTypePresencePossible(t, name, assumeTrue));
+                const filtered = filterType(type, t => isTypePresencePossible(t, name, assumeTrue));
+                // without `noUncheckedIndexedAccess` we can't narrow any further
+                // `{ [x: PropertyKey]: T }[prop]` already returns `T`
+                if (!compilerOptions.noUncheckedIndexedAccess) {
+                    return filtered;
+                }
+                return mapType(filtered, t => {
```
Comment (gabritto) [inline src/compiler/checker.ts:25509]: I think you should only do this if `assumeTrue` is true (maybe have the check above be `if (!compilerOptions.noUncheckedIndexedAccess || !assumeTrue) ...`).

Might be nice to add that to the tests as well, something like:
```ts
function f2(obj: Record<string, string>) {
  if ("test" in obj) {
    return obj.test
  }
  else {
    obj.test // Here it should be `string | undefined` still
  }
  return "default"
}
```

**Review by gabritto — CHANGES_REQUESTED** [review]
Small thing missing, but other than that, it looks good to me. Thanks 😊

**Comment by gabritto** [comment]
@ahejlsberg could you confirm we're ok with the solution proposed here of adding extra intersections with `Record<propName, propType>` during narrowing when `noUncheckedIndexedAccess` is true?

File: src/compiler/checker.ts (hunk 3 above)
Comment (Andarist) [reply src/compiler/checker.ts:25509]: great catch! thanks. I've pushed out the fix for this

**Comment by ahejlsberg** [comment]
First, I very much like the idea of narrowing indexed accesses guarded by `in` checks since we already do that for optional properties in `--exactOptionalPropertyTypes` mode. But, by that same reasoning, I think it should be done through the same mechanism as `--exactOptionalPropertyTypes`, i.e. use of the special `missingType` that indicates an `undefined` originating in a missing property. Specifically, we should use `missingType` as the `undefined` that is added when reading from an index signature and then rely on our existing CFA infrastructure to remove that `missingType` in code guarded by an `in` check. This would also get us support for narrowing through `obj.hasOwnProperty("foo")` checks, which is already in place for optional properties.

I realize that it is possible to get a similar effect using intersections, but it leads to more complex types and is oddly inconsistent with what we do for optional properties. And, really, index signatures are just another form of optional properties.

I will put up a PR in the next day or two to show what I mean. We may need to do a bit of optimizing of the logic that detects multiple kinds of `undefined` in unions and intersections, but we'll see what the performance suites reveal.

**Comment by Andarist** [comment]
Closing as superseded by https://github.com/microsoft/TypeScript/pull/51653

**Pull request closed without merging**